draft-ietf-privacypass-batched-tokens.md

```diff
@@ -319,13 +327,116 @@ struct {
+
+## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}
+
+Upon receipt of the request, the Issuer validates the following conditions:
+
+- The BatchTokenRequest contains a supported token_type registered with IANA.
+- The BatchTokenRequest contains token_requests with the same token_type, matching BatchTokenRequest.token_type.
+
+If any of these conditions is not met, the Issuer MUST return an HTTP 400 error
+to the client.
+For performance, an Issuer MAY return an HTTP 400 error if all token_requests are
+not targetted at the same cryptographic material.
+
+The Issuer then tries to deserialize the i-th element
+of BatchTokenRequest.token_requests using the protocol associated to BatchTokenRequest.token_type.
+If this fails for any of
+the BatchTokenRequest.token_requests values, the Issuer MUST return an HTTP 400
+error to the client. Otherwise, if the Issuer is willing to produce a token to
+the Client, the issuer creates a BatchTokenResponse structured as follows:
+
+
+~~~tls
+struct {
+  TokenResponse token_responses<0..2^16-1>;
+} BatchedTokenResponse
+~~~
+
+The structure fields are defined as follows:
+
+- "token_responses" are serialized TokenResponses, in network byte order. The number of token_responses, as a 2-octet integer, is prepended to the serialized TokenResponses This matches the number of incoming token_requests. In addition, the 2-octet integer length of each TokenResponse is prepended to the serialized TokenResponses.
+
+The Issuer generates an HTTP response with status code 200 whose content
+consists of TokenResponse, with the content type set as
+"application/private-token-batch-response".
+
+~~~
+HTTP/1.1 200 OK
+Content-Type: application/private-token-batch-response
+Content-Length: <Length of BatchTokenResponse>
+
+<Bytes containing the BatchTokenResponse>
+~~~
+
+An Issuer MAY accept streamed request, and MAY provide TokenResponse to the Client as they are ready.
+
+
+## Finalization {#arbitrary-finalization}
+
+The Client MUST validate that the TokenResponse.token_type matches the
+BatchedTokenRequest.token_type.
+
+The Client then tries to deserialize the i-th element
+of BatchTokenResponse.token_responses using the protocol associated to BatchTokenRequest.token_type.
+It then finalizes these tokens as specified in their respective protocols.
+
+If the FinalizeBatch function fails, the Client aborts the protocol. Token
+verification is unchanged.
+
 # Security considerations {#security-considerations}
 
+## Batched Privately Verifiable Tokens
+
 Implementors SHOULD be aware of the security considerations described in {{OPRF,
 Section 6.2.3}} and implement mitigation mechanisms. Application can mitigate
 this issue by limiting the number of clients and limiting the number of token
 requests per client per key.
 
+## Arbitrary Batched Verifiable Tokens
+
+Implementors SHOULD be aware of the inherent linear cost of this token type.
+Applications SHOULD place limits on the number of tokens per request.
```

Comment:
How do they impose this limit? I think they should have the ability to skip or ignore token requests if they're too many.

Reply:
I've updated this line to mention Issuer might skip token requests